alonzo/formal-spec/properties.tex

```diff
@@ -194,48 +275,308 @@ \section{Formal Properties}
-  If a transaction extends the UTxO, all its scripts validate, and
-  if it has a script that does not validate, it cannot extend the
-  UTxO.
-\item
-  \emph{Non-Token Minting.}
-  A valid transaction that does not mint tokens satisfies the
-  accounting property of the Shelley ledger where the type $\Coin$ is
-  replaced by $\Value$.
+For this property, we make some assumptions about things that are outside the scope of this specifications :
+
+\begin{itemize}
+  \item The Plutus script
+  interpreter is a pure function that receives only the arguments provided by the ledger when it is
+  invoked by calling the $\fun{runPLCScript}$ function. We assume that this is
+  also true in an implementation. In particular, the interpreter does not
+  obtain any system randomness, etc.
+
+  \item We do not need to check here that the hashes that are the keys of the
+  $\fun{txscripts}$ and $\fun{txdats}$ fields match the computed hashes of the scripts and
+  datum objects they index. This is because these hashes must be computed as part of
+  the deserialization of a transaction (see the CDDL specification),
+  instead of being transmitted as part of the transaction and then checked. We
+  assume these hashes are correct.
+
+  \item We assume that the consensus
+  function $\fun{epochInfoSlotToUTCTime}$ for converting a slot interval into a
+  system time interval is deterministic.
+
+  \item We assume that the two global
+  constants, $\EpochInfo$ and $\SystemStart$, which it also takes as parameters,
+  cannot change within the span of an interval for which $\fun{epochInfoSlotToUTCTime}$
+  is able to convert both endpoints to system time.
+\end{itemize}
+
+The assumptions above are implicit in the functional style of definitions in this
+specification, but they are worth pointing out for implementation.
+With these assumptions in mind, we can say that
+script evaluation is deterministic.
+
+We split this result into a lemma and its corollary :
+
+\begin{itemize}
+  \item First, we demonstrate that all the scripts and their arguments that are
+  collected for phase-2 validation are the same for two phase-1 valid transactions with the same body,
+  independent of the (necessarily valid) ledger state to which they are being applied
+
+  \item Second, we derive the corollary that under the same validity assumptions,
+  phase-2 validation results in the same outcome for both transactions
+\end{itemize}
+
+\begin{lemma}
+  \label{lem:inputs}
+  For all environments $e, e'$, transactions $tx, tx'$ and states $s, s', u, u'$ such that
+  $e$ and $s$ are subsets of fields of some valid chain state $c$, and
+  $e'$ and $s'$ are subsets of fields of some valid chain state $c'$,
+
+  \begin{equation*}
+    e\vdash s\trans{ledger}{tx}s', \\
+    e'\vdash u\trans{ledger}{tx'}u', \\
+    \txbody{tx} = \txbody{tx'}
+  \end{equation*}
+
+  The following holds :
+
+  \[\forall ~tp,~tp~\in~\fun{collectTwoPhaseScriptInputs} ~\field_{pp}(s)~\var{tx}~ \Utxo(s)\]
```

Comment:
I think using actual equality would be cleaner here

Reply:
that's true

Reply:
oh, hm - never mind. The return type of `collectTwoPhaseScriptInputs` is a list, and that the order of this list does not matter for evalScripts is discussed in the corollary that follows

Reply:
Right, I forgot that it's a list. Don't we have a `toSet` function that we already use? Then we could say that they are equal as sets. The issue I'm having with introducing a name here is that it sort of suggests that that element could be used elsewhere.

Reply:
oh yeah, toSet would make sense